KAFKA-16729: Support isolation level for share consumer

adixitconfluent opened the pull request:
### What
This PR adds the share group dynamic config `share.isolation.level`.
Until now, share groups only supported `READ_UNCOMMITTED` isolation
level type. With this PR, we aim to support `READ_COMMITTED` isolation
type to share groups.

### Tests
Added unit tests and integration tests to test the functionality.

### Approach
On a high level, the share partition leader is responsible for filtering
out batches that were a part of aborted transactions in `READ_COMMITTED`
isolation level. Broker acquires both the offsets containing control
records and offsets containing transactional records. The client is
expected to give an `ARCHIVED` acknowledgement for control record
batches.
On the broker, when fetch is being processed for `READ_COMMITTED`
isolation level, we immediately archive any offset/batch that was
aborted and only return the committed transactional record batches.

adixitconfluent committed:
Removed usage of partition max bytes from share fetch requests on the broker

adixitconfluent committed:
Removed spotless checkstyle failure

adixitconfluent committed:
Minor refactor

adixitconfluent committed:
Addressed Apoorv'r round 1 review comment

adixitconfluent committed:
Addressed Apoorv's round 1 review comments

adixitconfluent committed:
Addressed Apoorv's round 1 review comments

adixitconfluent committed:
Minor refactor

adixitconfluent committed:
Addresses Andrew's comments

adixitconfluent committed:
Branch creation

adixitconfluent committed:
Added isolationType to acquire method

adixitconfluent committed:
commit to save the porgress

adixitconfluent committed:
Added code + unit tests to add support for READ_COMMITTED isolation level

adixitconfluent committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16729

adixitconfluent committed:
Fixed tests and added share group dynamic config share.isolation.level

adixitconfluent committed:
Bug fix in iteration

adixitconfluent committed:
Fixed cases when an aborted transactions might already be a part of cached state + integration tests

adixitconfluent committed:
Cleanup

adixitconfluent committed:
Minor refactor

adixitconfluent committed:
Added more integration tests

adixitconfluent committed:
Minor comment change

adixitconfluent committed:
Fixed failing test

adixitconfluent committed:
Fixed flaky test

apoorvmittal10 commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -104,7 +104,8 @@ class KafkaApis(val requestChannel: RequestChannel,
                 val apiVersionManager: ApiVersionManager,
-                val clientMetricsManager: ClientMetricsManager
+                val clientMetricsManager: ClientMetricsManager,
+                val groupConfigManager: GroupConfigManager
```
Though it's fine to have another manager in KafkaApis but would it be better to have `groupConfigManager` in `SharePartitionManager`? KafkaApis can pass default IsolationLevel using `FetchIsolation.of(-1, GroupConfig.defaultShareIsolationLevel, true)` and SharePartitionManager can check for updated IsolationLevel using below code. The final isolation level can go in `ShareFetch` class? This approach require less plumbing in KafkaApis but adds a isolation level param in `ShareFetch` which might be different from `FetchParams`. What do you think?

```
val isolationLevel: IsolationLevel = IsolationLevel.forId(groupConfigManager.groupConfig(groupId).get().shareIsolationLevel().toByte)
          FetchIsolation.of(-1, isolationLevel, true)

adixitconfluent replied:
sounds good to me. I am happy doing it either way

apoorvmittal10 replied:
The other thing which you might want to check is that if we need to have correct Isolation Level in FetchParams from beginning, if somewhere used in ReplicaManager fetch.

adixitconfluent replied:
`isolation level` is utilzied in `DelayedShareFetch`, `ShareFetchUtils` and `SharePartition`, so we should be fine with this change moving to `SharePartitionManager`

apoorvmittal10 commented on server-common/src/main/java/org/apache/kafka/server/storage/log/FetchIsolation.java:
```diff
@@ -25,11 +25,11 @@ public enum FetchIsolation {
-    public static FetchIsolation of(int replicaId, IsolationLevel isolationLevel) {
-        if (!FetchRequest.isConsumer(replicaId)) {
+    public static FetchIsolation of(int replicaId, IsolationLevel isolationLevel, boolean isShareFetchRequest) {
+        if (!FetchRequest.isConsumer(replicaId) && !isShareFetchRequest) {
```
Hmmm, I am not sure if we should have `isShareFetchRequest` as a param. Shall there be any other replicaId for share consumers?

cc: @junrao 

```
    public static final int ORDINARY_CONSUMER_ID = -1;
    public static final int DEBUGGING_CONSUMER_ID = -2;
    public static final int FUTURE_LOCAL_REPLICA_ID = -3;

adixitconfluent replied:
I think if I just use `replicaId` as -1, that also works. But I wasn't too sure if that's the right approach. Hence, opted for adding a new variable

adixitconfluent replied:
I have made the code change to use `replicaId` as -1 and removed `isShareFetchRequest` param.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupConfig.java:
```diff
@@ -65,6 +67,13 @@ public final class GroupConfig extends AbstractConfig {
+    public static final String SHARE_ISOLATION_LEVEL_DOC = "Controls how to read records written transactionally. " +
+        "If set to \"read_committed\", the share group will only deliver transactional records which have been committed. " +
+        "If set to \"read_uncommitted\", the share group will return all messages, even transactional messages which have been aborted. " +
+        "Non-transactional records will be returned unconditionally in either mode.";
```
nit: would be good to align with share group configs above.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupConfig.java:
```diff
@@ -83,6 +92,8 @@ public final class GroupConfig extends AbstractConfig {
 
     public final int streamsNumStandbyReplicas;
 
+    public final int shareIsolationLevel;
```
KIP says `Valid values "read_committed"  and "read_uncommitted" (default)` so why do we have the int value?

adixitconfluent replied:
my bad. I misread it. Changed the code to reflect the right values.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -847,6 +859,7 @@ public ShareAcquiredRecords acquire(
             }
```
Don't we need the filtering here as well? Do we cover this scenario in unit test?

adixitconfluent replied:
we are doing the filtering after this point, so this case would be automatically covered.

apoorvmittal10 reviewed (commented):
Thanks for the PR. Some initial comments. I shall review the SharePartition class next.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,259 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+    }
+
+    private boolean containsAbortMarker(RecordBatch batch) {
+        lock.writeLock().lock();
```
Why this method code should have any lock?

adixitconfluent replied:
you're right, we don't need it.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,259 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+            if (!batchIterator.hasNext())
+                return false;
+
+            Record firstRecord = batchIterator.next();
+            return ControlRecordType.ABORT == ControlRecordType.parse(firstRecord.key());
```
Will it result in reading the records in memory? We should not be reading records in broker rather should rely on client to update. We should only read batch headers and not the records.

adixitconfluent replied:
I have replicated the code what is present in the regular consumer client for this case. When we do `batch.isControlBatch()`, we only read the batch header and not the entire record and we are doing a check on `record.key()` which involves reading the size of the key AFAIK. Not sure if there is a better way to do it.
cc - @AndrewJSchofield

AndrewJSchofield replied:
I think this is unavoidable. We are doing server-side filtering of records because otherwise we need to post-process the list of aborted transactions as part of dividing up fetched records for the set of consumers which are waiting for the data.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,259 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        }
+    }
+
+    private PriorityQueue<FetchResponseData.AbortedTransaction> abortedTransactionsHeap(List<FetchResponseData.AbortedTransaction> abortedTransactions) {
```
Can you please write the high level approach for the handling in PR description?

adixitconfluent replied:
highlighted the approach in PR description

apoorvmittal10 replied:
I meant regarding why do we need Priority Queue and those details. Anyways, I ll figure it out.

adixitconfluent replied:
Priority Queue makes our code more efficient, rather than iterating over all the offsets that we've already passed through.

apoorvmittal10 replied:
I am sure. My ask was to write the details down in description. Anyways it might be too much, I ll look into.

AndrewJSchofield replied:
This code essentially mirrors the way that aborted transactions are handled in `CompletedFetch.abortedTransactions`. If you look in detail at the way the information comes out of the replica manager, I think it's a good thing that the code being added to `SharePartition` is clearly doing the same thing. It would be easy to mess it up in odd cases by re-inventing the wheel.

apoorvmittal10 reviewed (changes requested):
Some high level comments around approach.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,259 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
         }
     }
 
+    private ShareAcquiredRecords filterAbortedTransactionalAcquiredRecords(FetchPartitionData fetchPartitionData, FetchIsolation isolationLevel, ShareAcquiredRecords shareAcquiredRecords) {
```
```suggestion
    private ShareAcquiredRecords maybeFilterAbortedTransactionalAcquiredRecords(FetchPartitionData fetchPartitionData, FetchIsolation isolationLevel, ShareAcquiredRecords shareAcquiredRecords) {
```

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,259 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+    private ShareAcquiredRecords filterAbortedTransactionalAcquiredRecords(FetchPartitionData fetchPartitionData, FetchIsolation isolationLevel, ShareAcquiredRecords shareAcquiredRecords) {
+        if (isolationLevel != FetchIsolation.TXN_COMMITTED)
+            return shareAcquiredRecords;
+        // When FetchIsolation.TXN_COMMITTED is used as isolation type by the share group, we need to filter any
```
Why not to have the below check in the same method and pass the aborted transactions to `filterAbortedTransactionalRecords` without optional. This can make sure that when it's required to filter then only call goes to further methods, and maybe method should have all these pre-checks.

```
if (abortedTransactions.isEmpty())
                return acquiredRecords;

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,259 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        return new ShareAcquiredRecords(result, acquiredCount);
+    }
+
+    private List<AcquiredRecords> filterAbortedTransactionalRecords(
```
Shouldn't the method name be consistent with earlier method `filterAbortedTransactionalAcquiredRecords` as we are sending the `acquired` records list here as well. Hence earlier method can have `maybe` prefix and this one as `filterAbortedTransactionalAcquiredRecords`.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,259 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+    private List<AcquiredRecords> filterAbortedTransactionalRecords(
+        Iterable<? extends RecordBatch> batches,
+        List<AcquiredRecords> acquiredRecords,
+        Optional<List<FetchResponseData.AbortedTransaction>> abortedTransactions
```
Are there any guarantees for `abortedTransactions` to be in some defined order with `firstOffset` for the RecordBatch?

adixitconfluent replied:
so far in the testing I have observed that the `abortedTransactions` contains all the first offset of all the fetched record batches that have been aborted in **ascending order** of the first offset.

adixitconfluent committed:
Addressed Apoorv's commet round 1  - part 1

adixitconfluent committed:
Addressed Apoorv's comment round 1  - part 2

adixitconfluent committed:
Addressed Apoorv's comment round 1  - part 3

adixitconfluent committed:
Fixed integration tests

adixitconfluent committed:
Simplified FetchIsolation

AndrewJSchofield commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupConfigTest.java:
```diff
@@ -247,6 +274,7 @@ private Properties createValidGroupConfig() {
         props.put(GroupConfig.STREAMS_SESSION_TIMEOUT_MS_CONFIG, "50000");
         props.put(GroupConfig.STREAMS_HEARTBEAT_INTERVAL_MS_CONFIG, "6000");
         props.put(GroupConfig.STREAMS_NUM_STANDBY_REPLICAS_CONFIG, "1");
+        props.put(GroupConfig.SHARE_ISOLATION_LEVEL_CONFIG, "read_uncommitted");
```
nit: I would tend to put all of the consumer configs, and then the share configs, and then the streams configs, just for neatness.

AndrewJSchofield commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupConfigTest.java:
```diff
@@ -67,6 +67,8 @@ public void testFromPropsInvalid() {
                 assertPropertyInvalid(name, "not_a_number", "1.0");
             } else if (GroupConfig.STREAMS_SESSION_TIMEOUT_MS_CONFIG.equals(name)) {
                 assertPropertyInvalid(name, "not_a_number", "1.0");
+            } else if (GroupConfig.SHARE_ISOLATION_LEVEL_CONFIG.equals(name)) {
```
nit: Please move up to the end of the other share configs.

AndrewJSchofield commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupConfigTest.java:
```diff
@@ -212,6 +237,7 @@ public void testFromPropsWithDefaultValue() {
         defaultValue.put(GroupConfig.STREAMS_HEARTBEAT_INTERVAL_MS_CONFIG, "10");
         defaultValue.put(GroupConfig.STREAMS_SESSION_TIMEOUT_MS_CONFIG, "2000");
         defaultValue.put(GroupConfig.STREAMS_NUM_STANDBY_REPLICAS_CONFIG, "1");
+        defaultValue.put(GroupConfig.SHARE_ISOLATION_LEVEL_CONFIG, "read_uncommitted");
```
nit: Move to the end of the other share group configs.

AndrewJSchofield commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -338,6 +338,7 @@ class KafkaApisTest extends Logging {
     cgConfigs.put(STREAMS_HEARTBEAT_INTERVAL_MS_CONFIG, GroupCoordinatorConfig.STREAMS_GROUP_HEARTBEAT_INTERVAL_MS_DEFAULT.toString)
     cgConfigs.put(STREAMS_SESSION_TIMEOUT_MS_CONFIG, GroupCoordinatorConfig.STREAMS_GROUP_SESSION_TIMEOUT_MS_DEFAULT.toString)
     cgConfigs.put(STREAMS_NUM_STANDBY_REPLICAS_CONFIG, GroupCoordinatorConfig.STREAMS_GROUP_NUM_STANDBY_REPLICAS_DEFAULT.toString)
+    cgConfigs.put(SHARE_ISOLATION_LEVEL_CONFIG, GroupConfig.SHARE_ISOLATION_LEVEL_DEFAULT)
```
nit: Please move to the end of the other share group configs.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -683,6 +692,7 @@ public long nextFetchOffset() {
      *                           if the records are already part of the same fetch batch.
      * @param fetchOffset        The fetch offset for which the records are fetched.
      * @param fetchPartitionData The fetched records for the share partition.
+     * @param isolationType      The isolation level for the share fetch request.
```
nit: It's `isolationLevel` not `isolationType` I think. Even the comment agrees with me ;0)

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartitionManager.java:
```diff
@@ -267,8 +268,20 @@ public CompletableFuture<Map<TopicIdPartition, PartitionData>> fetchMessages(
+                fetchParams.maxWaitMs,
+                fetchParams.minBytes,
+                fetchParams.maxBytes,
+                FetchIsolation.of(-1, groupConfigManager.groupConfig(groupId).get().shareIsolationLevel()),
```
This `-1` can be `FetchRequest.CONSUMER_REPLICA_ID`.

adixitconfluent replied:
@AndrewJSchofield, earlier I had thought to use `isShareFetchRequest` as a parameter within `FetchIsolation.of`. That idea did not go well as per discussion https://github.com/apache/kafka/pull/19261#discussion_r2012532595. Do you think I should add a new variable like below in `FetchRequest.java` or `ShareFetchRequest.java`?
`public static final int SHARE_CONSUMER_ID = -4;`

AndrewJSchofield replied:
I think what I object to is using a bare `-1` when there is a specific constant name for `-1` already in existence. I believe that this parameter is strictly the replica ID, not really a consumer ID, but it's all a bit mixed up in `FetchRequest`. We certainly could introduce a `SHARE_CONSUMER_REPLICA_ID` I suppose, but if it's really just used for this situation, masquerading as a regular consumer with `CONSUMER_REPLICA_ID` seems fine to me.

AndrewJSchofield commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -3239,7 +3239,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         fetchMinBytes,
         fetchMaxBytes,
-        FetchIsolation.HIGH_WATERMARK,
+        FetchIsolation.of(-1, GroupConfig.defaultShareIsolationLevel),
```
This `-1` can be `FetchRequest.CONSUMER_REPLICA_ID`.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,243 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+    private ShareAcquiredRecords maybeFilterAbortedTransactionalAcquiredRecords(FetchPartitionData fetchPartitionData, FetchIsolation isolationLevel, ShareAcquiredRecords shareAcquiredRecords) {
+        if (isolationLevel != FetchIsolation.TXN_COMMITTED || fetchPartitionData.abortedTransactions.isEmpty())
+            return shareAcquiredRecords;
+        // When FetchIsolation.TXN_COMMITTED is used as isolation type by the share group, we need to filter any
```
Isolation level, not type, please.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,243 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
         }
     }
 
+    private ShareAcquiredRecords maybeFilterAbortedTransactionalAcquiredRecords(FetchPartitionData fetchPartitionData, FetchIsolation isolationLevel, ShareAcquiredRecords shareAcquiredRecords) {
```
nit: This source file tends to put the parameter list over multiple lines when it gets this long.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,243 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+    }
+
+    private PriorityQueue<FetchResponseData.AbortedTransaction> abortedTransactionsHeap(List<FetchResponseData.AbortedTransaction> abortedTransactions) {
+        if (abortedTransactions == null || abortedTransactions.isEmpty())
```
I think we know that neither of these conditions is true. I'd prefer to take advantage of that knowledge so that the caller does not need to handle the null return here with its own conditional logic.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,243 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        }
+    }
+
+    private PriorityQueue<FetchResponseData.AbortedTransaction> abortedTransactionsHeap(List<FetchResponseData.AbortedTransaction> abortedTransactions) {
```
The name `abortedTransactionsHeap` seems very clumsy to me. What you are doing is ensuring that the returned list is ordered by increasing first offset. So, "sorted" or "ordered" might be useful in the name.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. I have made some comments from my initial review. Still working on `SharePartition`.

adixitconfluent committed:
Addressed Andrew's round 1 comments

adixitconfluent committed:
Addressed Andrew's round 1 remaining comments

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupConfig.java:
```diff
@@ -290,6 +309,10 @@ public static ShareGroupAutoOffsetResetStrategy defaultShareAutoOffsetReset() {
         return ShareGroupAutoOffsetResetStrategy.fromString(SHARE_AUTO_OFFSET_RESET_DEFAULT);
     }
 
+    public static IsolationLevel defaultShareIsolationLevel() {
```
Please write javadoc as being followed for other public methods in the class.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupConfig.java:
```diff
@@ -352,4 +375,15 @@ public int streamsHeartbeatIntervalMs() {
         return streamsNumStandbyReplicas;
     }
+
+    public IsolationLevel shareIsolationLevel() {
```
Please write javadoc as being followed for other public methods in the class.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,241 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+
+        Record firstRecord = batchIterator.next();
+        return ControlRecordType.ABORT == ControlRecordType.parse(firstRecord.key());
+
```
nit: remove blank line

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,241 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+            return shareAcquiredRecords;
+        // When FetchIsolation.TXN_COMMITTED is used as isolation level by the share group, we need to filter any
+        // transactions that were aborted/did not commit due to timeout.
+        List<AcquiredRecords> result = filterAbortedTransactionalAcquiredRecords(fetchPartitionData.records.batches(), shareAcquiredRecords.acquiredRecords(), fetchPartitionData.abortedTransactions.get());
```
nit: break the line in 2 lines please.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,241 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        }
+    }
+
+    private NavigableMap<Long, InFlightBatch> fetchSubMap(RecordBatch recordBatch) {
```
Please write method comments for the method. Also should the name be `fetchSubMapOrException`?

AndrewJSchofield replied:
No, I don't think it should.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,241 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+            if (floorOffset == null) {
+                log.debug("Fetched batch record {} not found for share partition: {}-{}", recordBatch, groupId,
+                    topicIdPartition);
+                throw new InvalidRecordStateException(
+                    "Batch record not found. The request batch offsets are not found in the cache.");
```
Should it be `InvalidRecordStateException` when some offset is not found?

adixitconfluent replied:
Do you think it should be `IllegalStateException`?

apoorvmittal10 replied:
`IllegalStateException` might be better as this condition should not hit.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,241 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+    private NavigableMap<Long, InFlightBatch> fetchSubMap(RecordBatch recordBatch) {
+        lock.writeLock().lock();
+        try {
+            Map.Entry<Long, InFlightBatch> floorOffset = cachedState.floorEntry(recordBatch.baseOffset());
```
```suggestion
            Map.Entry<Long, InFlightBatch> floorEntry = cachedState.floorEntry(recordBatch.baseOffset());
```

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,241 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+    }
+
+    private NavigableMap<Long, InFlightBatch> fetchSubMap(RecordBatch recordBatch) {
+        lock.writeLock().lock();
```
Should it be readLock?

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,241 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        Iterable<? extends RecordBatch> batches,
+        List<FetchResponseData.AbortedTransaction> abortedTransactions
+    ) {
+        lock.writeLock().lock();
```
What operations in the method requires lock?

adixitconfluent replied:
you're right, we don't require it.

adixitconfluent replied:
you're right, we don't need it.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,241 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        }
+    }
+
+    private void archiveAcquiredBatchRecords(NavigableMap<Long, InFlightBatch> subMap, RecordBatch recordBatch) {
```
Should these `archive` methods be aligned together with `archiveAvailableRecords`.  Moreover seems the method is similar to `archiveAvailableRecords` itself, can't there be common method as the required batch/offset state can be passed as method argument?

adixitconfluent replied:
`archiveAvailableRecords` is quite different than `archiveAcquiredBatchRecords` in several ways- 
1. `archiveAcquiredBatchRecords` doesn't need a return type unlike `archiveAvailableRecords`
2. `archiveAvailableRecords` requires more checks than what are needed.

Hence, I prefer keeping them as separate rather than adding more if else statements.

apoorvmittal10 replied:
Return types can be ignored. I re-compared both and still thinks that we should merge. The comparison of startOffset and endOffset seems relevant in both methods and we should have them. Let me know if I am missing something.

```
private void archiveAcquiredBatchRecords(NavigableMap<Long, InFlightBatch> subMap, RecordBatch recordBatch) {
     archiveAvailableRecords(recordBatch.firstOffset(), recordBatch.lastOffset() + 1, subMap, Acquired);
```

```
private boolean archiveAvailableRecords(long startOffset, long endOffset, NavigableMap<Long, InFlightBatch> map, RecordState initialState)
```

```
move archiveAvailableRecords => archiveRecords
```

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,241 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        }
+    }
+
+    private void archivePerOffsetAcquiredBatchRecords(InFlightBatch inFlightBatch, long startOffsetToArchive, long endOffsetToArchive) {
```
Seems similar to `archivePerOffsetBatchRecords`, can't there be common code?

adixitconfluent replied:
I can introduce a method parameter `RecordState initialState` and make it a common function. Done this in my latest commit.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2497,241 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        }
+    }
+
+    private void archiveCompleteAcquiredBatch(InFlightBatch inFlightBatch) {
```
Seems similar to `archiveCompleteBatch`.

adixitconfluent replied:
I can introduce a method parameter `RecordState initialState` and make it a common function. Done this in my latest commit.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartitionManager.java:
```diff
@@ -267,8 +269,20 @@ public CompletableFuture<Map<TopicIdPartition, PartitionData>> fetchMessages(
         CompletableFuture<Map<TopicIdPartition, PartitionData>> future = new CompletableFuture<>();
-        processShareFetch(new ShareFetch(fetchParams, groupId, memberId, future, rotatedTopicIdPartitions, batchSize, maxFetchRecords, brokerTopicStats));
-
+        if (groupConfigManager.groupConfig(groupId).isEmpty()) {
+            processShareFetch(new ShareFetch(fetchParams, groupId, memberId, future, rotatedTopicIdPartitions, batchSize, maxFetchRecords, brokerTopicStats));
+        } else {
+            FetchParams updatedFetchParams = new FetchParams(
+                fetchParams.replicaId,
+                fetchParams.replicaEpoch,
+                fetchParams.maxWaitMs,
+                fetchParams.minBytes,
+                fetchParams.maxBytes,
+                FetchIsolation.of(FetchRequest.CONSUMER_REPLICA_ID, groupConfigManager.groupConfig(groupId).get().shareIsolationLevel()),
+                fetchParams.clientMetadata
+            );
+            processShareFetch(new ShareFetch(updatedFetchParams, groupId, memberId, future, rotatedTopicIdPartitions, batchSize, maxFetchRecords, brokerTopicStats));
+        }
```
Please have a method `maybUpdateFetchParams` which shall return updated fetchParams and use it here.
```
FetchParams updatedFetchParams = maybeUpdateFetchParams(fetchParams);
processShareFetch(new ShareFetch(updatedFetchParams, groupId, memberId, future, rotatedTopicIdPartitions, batchSize, maxFetchRecords, brokerTopicStats));
```
```
private FetchParams  maybUpdateFetchParams(FetchParams fetchParams) {
if (groupConfigManager.groupConfig(groupId).isEmpty()) {
            return fetchParams;
}
 FetchParams updatedFetchParams = new FetchParams(
                fetchParams.replicaId,
                fetchParams.replicaEpoch,
                fetchParams.maxWaitMs,
                fetchParams.minBytes,
                fetchParams.maxBytes,
                FetchIsolation.of(FetchRequest.CONSUMER_REPLICA_ID, groupConfigManager.groupConfig(groupId).get().shareIsolationLevel()),
                fetchParams.clientMetadata
            );
return updatedFetchParams;
}

apoorvmittal10 reviewed (changes requested):
Thanks for making the changes, some comments.

adixitconfluent committed:
Adressed Apoorv's review comments round 2 - part 1

adixitconfluent committed:
Adressed Apoorv's review comments round 2 - part 2

apoorvmittal10 commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -3239,7 +3239,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         fetchMinBytes,
         fetchMaxBytes,
-        FetchIsolation.HIGH_WATERMARK,
+        FetchIsolation.of(FetchRequest.CONSUMER_REPLICA_ID, GroupConfig.defaultShareIsolationLevel),
```
Just FYI: there was a need to have `groupConfigManager` in KafkaApis as the final response is creted in KafkaApis hence I added it in following PR: https://github.com/apache/kafka/pull/19334/files. Once merged, we can move the code back here from SharePartitionManager. Sorry for the trouble.

adixitconfluent replied:
I have moved back the resolution of fetch isolation level to `KafkaApis` from `SharePartitionManager`

apoorvmittal10 replied:
Thanks 👍

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2504,205 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+            if (currentBatch.hasProducerId()) {
+                // remove from the aborted transactions queue, all aborted transactions which have begun before the
+                // current batch's last offset and add the associated producerIds to the aborted producer set.
+                while (!orderedAbortedTransactions.isEmpty() && orderedAbortedTransactions.peek().firstOffset() <= currentBatch.lastOffset()) {
+                    FetchResponseData.AbortedTransaction abortedTransaction = orderedAbortedTransactions.poll();
+                    abortedProducerIds.add(abortedTransaction.producerId());
+                }
+                long producerId = currentBatch.producerId();
+                if (containsAbortMarker(currentBatch)) {
+                    abortedProducerIds.remove(producerId);
+                } else if (isBatchAborted(currentBatch, abortedProducerIds)) {
+                    log.debug("Skipping aborted record batch for share partition: {}-{} with producerId {} and " +
+                        "offsets {} to {}", groupId, topicIdPartition, producerId, currentBatch.baseOffset(), currentBatch.lastOffset());
+                    recordsToArchive.add(currentBatch);
```
The code is not covered in tests, can we please add scenarios which can test the code.

adixitconfluent replied:
I have added unit tests for different scenarios corresponding to this code.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2504,205 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        return orderedAbortedTransactions;
+    }
+
+    private boolean isBatchAborted(RecordBatch batch, Set<Long> abortedProducerIds) {
+        return batch.isTransactional() && abortedProducerIds.contains(batch.producerId());
+    }
+
+    private boolean containsAbortMarker(RecordBatch batch) {
+        if (!batch.isControlBatch())
+            return false;
+
+        Iterator<Record> batchIterator = batch.iterator();
+        if (!batchIterator.hasNext())
+            return false;
+
+        Record firstRecord = batchIterator.next();
+        return ControlRecordType.ABORT == ControlRecordType.parse(firstRecord.key());
+    }
```
Are there scenrios which tests these methods?

adixitconfluent replied:
The functions `orderedAbortedTransactions`, `isBatchAborted`, `containsAbortMarker` are mirrors of what happens in `CompletedFetch` class. You won't find any unit tests for these functions there. That's why I've avoided them here as well. I am testing the logic that comes after we get aborted transactional records provided by function `fetchAbortedTransactionRecordBatches`. You'll be finding mocks for this function in `SharePartitionTest`
cc- @AndrewJSchofield

apoorvmittal10 replied:
> You won't find any unit tests for these functions there. That's why I've avoided them here as well.

It must be a miss there that doesn't mean we should avoid them too. There should be tests for the code we are writing to make sure it works as expected.

adixitconfluent replied:
I have added unit tests for different scenarios corresponding to this code.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1241,7 +1254,8 @@ private boolean archiveAvailableRecords(long startOffset, long endOffset, Naviga
                                                  long startOffsetToArchive,
-                                                 long endOffsetToArchive) {
+                                                 long endOffsetToArchive,
+                                                 RecordState initialState) {
```
```suggestion
                                                 RecordState initialState
     ) {
```

apoorvmittal10 replied:
nit: seems missed.

apoorvmittal10 reviewed (changes requested):
Mostly LGTM, some comments. Are all scenarios covered in SharePartition unit tests?

adixitconfluent commented:
> Are all scenarios covered in SharePartition unit tests?

Hi @apoorvmittal10, I have added the unit tests for different cases of acquire (cached state can be empty/acquire full batch/acquire subset batch). I think a better way to test the added code was through integration tests, hence there I have added plenty in `ShareConsumerTest` since I can mimic transactions and there the code to process aborted transactions is also being tested. Please let me know if this doesn't look enough and I should add more?
cc - @AndrewJSchofield

adixitconfluent committed:
Adressed Apoorv's review comments round 3 - part 1

adixitconfluent committed:
Added more unit tests as per Apoorv's comments

adixitconfluent committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16729

adixitconfluent committed:
Fix failing integration tests because of the merge

adixitconfluent committed:
Resolved the group isolationLevel in KafkaApis

adixitconfluent commented:
Hi @AndrewJSchofield @apoorvmittal10 , the PR is ready for review. Sorry for a force-push, I had to do it since my git history had become corrupt due to a problematic merge. Because of that, there are a lot more labels to this PR than required. Can you please remove the extra labels and re-review my PR?

smjn commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/ShareConsumerTest.java:
```diff
@@ -2090,6 +2091,307 @@ public void testComplexShareConsumer() throws Exception {
+                produceAbortedTransaction(transactionalProducer, "Message 2");
+
+                // We will not receive any records since the transaction was aborted.
+                records = waitedPoll(shareConsumer, 2500L, 0);
```
This might not be enough. Ideally, we should call `waitForAssignment` to make sure something has been assigned as well as 0 records.

AndrewJSchofield commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -3199,13 +3199,21 @@ class KafkaApis(val requestChannel: RequestChannel,
         fetchMinBytes,
         fetchMaxBytes,
-        FetchIsolation.HIGH_WATERMARK,
+        fetchIsolation(),
```
Maybe this could be simplified to something like
```FetchIsolation.of(FetchRequest.CONSUMER_REPLICA_ID, groupConfigManager.groupConfig(groupId).map(config => config.shareIsolationLevel()).orElse(GroupConfig.defaultShareIsolationLevel))```

AndrewJSchofield replied:
Although "simplified" might be stretching the point.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupConfig.java:
```diff
@@ -137,7 +149,13 @@ public final class GroupConfig extends AbstractConfig {
             MEDIUM,
-            GroupCoordinatorConfig.STREAMS_GROUP_NUM_STANDBY_REPLICAS_DOC);
+            GroupCoordinatorConfig.STREAMS_GROUP_NUM_STANDBY_REPLICAS_DOC)
+        .define(SHARE_ISOLATION_LEVEL_CONFIG,
```
Please put this config adjacent to the other share group configs above.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1206,7 +1219,7 @@ private boolean archiveAvailableRecordsOnLsoMovement(long logStartOffset) {
      * @param map The map containing the in-flight records.
      * @return A boolean which indicates whether any record is archived or not.
```
nit: Add `initialState` to the javadoc comment.

AndrewJSchofield replied:
And correct the description so it doesn't just return to the available records.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1246,7 +1259,9 @@ private boolean archiveAvailableRecords(long startOffset, long endOffset, Naviga
-                                                 long endOffsetToArchive) {
+                                                 long endOffsetToArchive,
+                                                 RecordState initialState
+    ) {
```
nit: This source file tends to have the `) {` on the end of the parameter line.

adixitconfluent replied:
I made the change as per this review comment https://github.com/apache/kafka/pull/19261#discussion_r2022398987 but I agree that this file has the convention to use  `) { ` on the end of the parameter line.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,174 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+    ) {
+        if (isolationLevel != FetchIsolation.TXN_COMMITTED || fetchPartitionData.abortedTransactions.isEmpty() || fetchPartitionData.abortedTransactions.get().isEmpty())
+            return shareAcquiredRecords;
+        // When FetchIsolation.TXN_COMMITTED is used as isolation level by the share group, we need to filter any
```
nit: I'd add a blank line to make it a bit more readable at a glance. The general case is to return on line 2514.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,174 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+            for (RecordBatch recordBatch : recordsToArchive) {
+                // Archive the offsets/batches in the cached state.
+                NavigableMap<Long, InFlightBatch> subMap = fetchSubMapOrException(recordBatch);
+                archiveAcquiredBatchRecords(subMap, recordBatch);
```
Given that this method is only really a single line of code, I would remove the method and just have the line of code here.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,174 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+     * @param recordBatch The record batch for which we want to find the sub map.
+     * @return the sub map containing all the offset details.
+     */
+    private NavigableMap<Long, InFlightBatch> fetchSubMapOrException(RecordBatch recordBatch) {
```
I think this style is a bit strange. If I understand correctly, we never expect to see the exception. That would be a bug. So, I would just call it `getSubMap` or `getCachedSubMap` or similar. Yes, it throws an exception if it's called in an invalid state, but then the share-partition would be messed up already.

adixitconfluent replied:
I agree to your point. Earlier I had accepted the suggestion as per review comment https://github.com/apache/kafka/pull/19261#discussion_r2021438959, but I'll change it back to `fetchSubMap`

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,174 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+                                newAcquiredRecords.add(new AcquiredRecords()
+                                    .setFirstOffset(temp.firstOffset())
+                                    .setLastOffset(recordBatch.baseOffset() - 1)
+                                    .setDeliveryCount((short) 1));
```
Wouldn't this be `temp.deliveryCount()`?

adixitconfluent replied:
good catch, yes

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,174 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+                                newAcquiredRecords.add(new AcquiredRecords()
+                                    .setFirstOffset(recordBatch.lastOffset() + 1)
+                                    .setLastOffset(temp.lastOffset())
+                                    .setDeliveryCount((short) 1));
```
Also here.

AndrewJSchofield commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -116,6 +124,7 @@ public class SharePartitionTest {
     private static final int DEFAULT_FETCH_OFFSET = 0;
     private static final int MAX_FETCH_RECORDS = Integer.MAX_VALUE;
     private static final byte ACKNOWLEDGE_TYPE_GAP_ID = 0;
+    private static final FetchIsolation FETCH_ISOLATION = FetchIsolation.HIGH_WATERMARK;
```
This constant name really tells us the type, not the value. I wonder if `FETCH_ISOLATION_HWM` might be more instructive.

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,174 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        try {
+            List<AcquiredRecords> result = new ArrayList<>();
+
+            for (AcquiredRecords acquiredRecord : acquiredRecords) {
```
It is a bit odd having the loop variable of type `AcquiredRecords` called `acquiredRecord`, and the list of `AcquiredRecords` called `acquiredRecords`.

adixitconfluent replied:
I have renamed `acquiredRecords` to `acquiredRecordsList` and `acquiredRecord` to `acquiredRecords`

AndrewJSchofield commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,174 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+            for (AcquiredRecords acquiredRecord : acquiredRecords) {
+                List<AcquiredRecords> tempAcquiredRecords = new ArrayList<>();
+                tempAcquiredRecords.add(acquiredRecord);
+                for (RecordBatch recordBatch : recordsToArchive) {
```
If you renamed `recordsToArchive` as `batchesToArchive`, and then renamed `recordBatch` to `batchToArchive`, I think that would help with the readability.

AndrewJSchofield reviewed (changes requested):
Thanks for the updates. In general, I'm happy with the PR, with the exception of the `filterRecordBatchesFromAcquiredRecords` method. It needs much more comprehensive commenting and attention to naming. For example, it's not clear whether the input parameters are sorted and so on. This makes it difficult to work out the algorithm that it is using.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,174 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+                tempAcquiredRecords.add(acquiredRecord);
+                for (RecordBatch recordBatch : recordsToArchive) {
+                    List<AcquiredRecords> newAcquiredRecords = new ArrayList<>();
+                    for (AcquiredRecords temp : tempAcquiredRecords) {
```
This is probably ok, but we could have just iterated recordsToArchive and tempAcquiredRecords once to do a merge instead of double loop.

adixitconfluent replied:
I have optimized the code to avoid the double loop and iterate once.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,174 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        List<AcquiredRecords> acquiredRecords,
+        List<FetchResponseData.AbortedTransaction> abortedTransactions
+    ) {
+        lock.writeLock().lock();
```
Do we need to lock here? The only caller is `acquire()` and already holds the lock. Ditto for the locking in `archiveRecords()`

adixitconfluent replied:
agreed, we don't need it for `filterAbortedTransactionalAcquiredRecords`

adixitconfluent replied:
I think `archiveRecords()` needs a lock because `inFlightBatch.maybeInitializeOffsetStateUpdate()` is called which can change the cachedState

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,174 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        List<AcquiredRecords> acquiredRecords,
+        List<RecordBatch> recordsToArchive
+    ) {
+        lock.writeLock().lock();
```
Do we need to lock here since the caller already holds the lock?

adixitconfluent replied:
agreed, we don't need it.

junrao reviewed (commented):
@adixitconfluent : Thanks for the PR. Made a pass of non-testing files. Just a few minor comments.

adixitconfluent committed:
Adressed Andrew's review comments round 2 - part 1

adixitconfluent committed:
Adressed Andrew's review comments round 2 - remaining comments

adixitconfluent commented:
> Thanks for the updates. In general, I'm happy with the PR, with the exception of the `filterRecordBatchesFromAcquiredRecords` method. It needs much more comprehensive commenting and attention to naming. For example, it's not clear whether the input parameters are sorted and so on. This makes it difficult to work out the algorithm that it is using.

Hi @AndrewJSchofield, thanks for the feedback. I have made naming changes as per your review comments, added the javadoc over the method and more comments in this method explaining my approach. Please let me know if this is explanatory enough or I need more changes. I have also optimized the code to avoid double loops. Thanks!

adixitconfluent committed:
Adressed Jun's review comments round 1 - part 1

adixitconfluent committed:
Optimized function filterRecordBatchesFromAcquiredRecords

adixitconfluent committed:
Added more tests for filterRecordBatchesFromAcquiredRecords

AndrewJSchofield commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -1175,7 +1185,8 @@ public void testAcquireWithMaxFetchRecords() {
             DEFAULT_FETCH_OFFSET,
-            fetchPartitionData(records)),
+            fetchPartitionData(records),
+                FETCH_ISOLATION_HWM),
```
nit: indentation

AndrewJSchofield commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -1212,7 +1223,8 @@ public void testAcquireWithMultipleBatchesAndMaxFetchRecords() throws Interrupte
             DEFAULT_FETCH_OFFSET,
-            fetchPartitionData(records, 10)),
+            fetchPartitionData(records, 10),
+                FETCH_ISOLATION_HWM),
```
nit: indentation

AndrewJSchofield commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -1308,7 +1321,8 @@ public void testAcquireWithBatchSizeAndSingleBatch() {
             DEFAULT_FETCH_OFFSET,
-            fetchPartitionData(records)),
+            fetchPartitionData(records),
+                FETCH_ISOLATION_HWM),
```
nit: indentation

AndrewJSchofield commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -1339,7 +1353,8 @@ public void testAcquireWithBatchSizeAndMultipleBatches() {
             DEFAULT_FETCH_OFFSET,
-            fetchPartitionData(records)),
+            fetchPartitionData(records),
+                FETCH_ISOLATION_HWM),
```
nit: indentation

AndrewJSchofield commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -6671,15 +6693,441 @@ private String assertionFailedMessage(SharePartition sharePartition, Map<Long, L
+            sharePartition.filterRecordBatchesFromAcquiredRecords(acquiredRecords1, recordBatches1));
+
+        List<AcquiredRecords> acquiredRecords2 = List.of(
+            new AcquiredRecords().setFirstOffset(1).setLastOffset(4).setDeliveryCount((short) 1),
```
Let's put some variation in the delivery count here.

AndrewJSchofield reviewed (changes requested):
Thanks for the updates. I am happy with the code now and have just a few comments on the tests. It's getting there.

adixitconfluent committed:
Adressed Andrew's review comments round 3

AndrewJSchofield reviewed (approved):
lgtm

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,182 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+    ) {
+        List<AcquiredRecords> result = new ArrayList<>();
+        for (AcquiredRecords acquiredRecords : acquiredRecordsList) {
+            List<AcquiredRecords> tempAcquiredRecordsList = new ArrayList<>();
```
Hmm, I don't quite understand this. tempAcquiredRecordsList seems to always contain a single AcquiredRecords. Why does it need to be a list?

AndrewJSchofield replied:
It's because the next iteration can divide the single AcquiredRecords, and then loop again. It's this iterative refinement that necessitates the list. It is true that the list is often only a single element.

junrao replied:
Got it. The usage of `tempAcquiredRecordsIndex` still seems unintuitive.

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2484,6 +2505,182 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+        for (AcquiredRecords acquiredRecords : acquiredRecordsList) {
+            List<AcquiredRecords> tempAcquiredRecordsList = new ArrayList<>();
+            tempAcquiredRecordsList.add(acquiredRecords);
+            for (RecordBatch batchToArchive : batchesToArchive) {
```
This might be ok for now. But for each acquiredRecords, we are still iterating every batchToArchive. Since both acquiredRecordsList and batchesToArchive are in offset order, we could just iterating them together in one pass?

adixitconfluent replied:
I don't think a single pass will work. Because there could be a case that each element of `acquiredRecordsList` breaks up as we pass the `batchesToArchive`, and in doing so we might need to refine that element into parts. These parts need iteration to refine them further, if required. Hence, I think we need the double pass here.

junrao replied:
I was thinking of sth like the following. Will it work?
```
        List<AcquiredRecords> result = new ArrayList<>();
        Iterator<AcquiredRecords> acquiredRecordsListIter = acquiredRecordsList.iterator();
        Iterator<RecordBatch> batchesToArchiveIterator = batchesToArchive.iterator();
        if (!batchesToArchiveIterator.hasNext())
            return acquiredRecordsList;
        RecordBatch batchToArchive = batchesToArchiveIterator.next();
        AcquiredRecords unresolvedAcquiredRecords = null;

        while (unresolvedAcquiredRecords != null || acquiredRecordsListIter.hasNext()) {
            if (unresolvedAcquiredRecords == null)
                unresolvedAcquiredRecords = acquiredRecordsListIter.next();

            long unresolvedFirstOffset = unresolvedAcquiredRecords.firstOffset();
            long unresolvedLastOffset = unresolvedAcquiredRecords.lastOffset();
            short unresolvedDeliveryCount = unresolvedAcquiredRecords.deliveryCount();

            if (batchToArchive == null) {
                result.add(unresolvedAcquiredRecords);
                unresolvedAcquiredRecords = null;
                continue;
            }

            // Non-overlap check - acquired records offsets lie before the batchToArchive offsets. No need to filter out the offsets in such a scenario.
            if (unresolvedLastOffset < batchToArchive.baseOffset()) {
                // Offsets in tempAcquiredRecordsList element do not overlap with batchToArchive, hence it should not get filtered out.
                result.add(unresolvedAcquiredRecords);
                unresolvedAcquiredRecords = null;
            }

            // Overlap check - acquired records offsets overlap with the record batch offsets. We need to filter out the overlapping
            // offsets in such a scenario.
            if (unresolvedFirstOffset <= batchToArchive.lastOffset() &&
                    unresolvedLastOffset >= batchToArchive.baseOffset()) {
                unresolvedAcquiredRecords = null;
                // Split the acquired record into parts - before and after the overlapping record batch.
                if (unresolvedFirstOffset < batchToArchive.baseOffset()) {
                    // The offsets in tempAcquiredRecords that are present before batchToArchive's baseOffset should not get filtered out.
                    result.add(new AcquiredRecords()
                            .setFirstOffset(unresolvedFirstOffset)
                            .setLastOffset(batchToArchive.baseOffset() - 1)
                            .setDeliveryCount(unresolvedDeliveryCount));
                }
                if (unresolvedLastOffset > batchToArchive.lastOffset()) {
                    // The offsets in tempAcquiredRecords that are present after batchToArchive's lastOffset should not get filtered out.
                    unresolvedAcquiredRecords = new AcquiredRecords()
                            .setFirstOffset(batchToArchive.lastOffset() + 1)
                            .setLastOffset(unresolvedLastOffset)
                            .setDeliveryCount(unresolvedDeliveryCount);
                }
            }

            if (unresolvedLastOffset >= batchToArchive.lastOffset()) {
                if (batchesToArchiveIterator.hasNext())
                    batchToArchive = batchesToArchiveIterator.next();
                else
                    batchToArchive = null;
            }
        }

        return result;
```

adixitconfluent replied:
Hi @junrao , thanks for this approach. I agree that variables in this approach are more intuitive and it looks more readable. I have taken your changes and made a few changes to the code comments (in line with your code).

junrao reviewed (commented):
@adixitconfluent : Thanks for the updated PR. A couple of more comments.

adixitconfluent committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16729 to resolve conflicts with trunk

junrao reviewed (commented):
@adixitconfluent : Thanks for rebasing. One more followup comment.

adixitconfluent committed:
Adressed Jun's review comments to refactor filterRecordBatchesFromAcquiredRecords

junrao commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2488,6 +2509,190 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+                }
+            }
+
+            if (unresolvedLastOffset >= batchToArchive.lastOffset()) {
```
Could we add a comment for this case?

junrao commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -6671,15 +6693,441 @@ private String assertionFailedMessage(SharePartition sharePartition, Map<Long, L
+
+        assertEquals(List.of(
+            new AcquiredRecords().setFirstOffset(15).setLastOffset(19).setDeliveryCount((short) 1),
+            new AcquiredRecords().setFirstOffset(20).setLastOffset(49).setDeliveryCount((short) 1),
```
Why do we set the last offset to 49 when the original batch ends at offset 34?

adixitconfluent replied:
This is because there is a gap of records from 35-49 in the original records to be acquired. This gap is added to this batch since we rely on the client to inform the broker about these natural gaps in the partition log.
```
memoryRecordsBuilder(buffer, 5, 10).close(); // batch from 10-14
memoryRecordsBuilder(buffer, 5, 15).close(); // batch from 15-19
memoryRecordsBuilder(buffer, 15, 20).close(); // batch from 20-34
memoryRecordsBuilder(buffer, 8, 50).close();  // batch from 50-57
```

junrao commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -6671,15 +6693,441 @@ private String assertionFailedMessage(SharePartition sharePartition, Map<Long, L
+        when(recordBatch.isControlBatch()).thenReturn(false);
+        assertFalse(sharePartition.containsAbortMarker(recordBatch));
+
+        // Record batch is a control batch but doesn't contain any records..
```
extra .

junrao commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -6671,15 +6693,441 @@ private String assertionFailedMessage(SharePartition sharePartition, Map<Long, L
+        buffer.flip();
+        Records records = MemoryRecords.readableRecords(buffer);
+
+        // Case 1 - Aborted transactions does not contain the record batch from 4-5 with producer id 2.
```
Hmm, batch 2-3 have produce id 2.

adixitconfluent replied:
Corrected the comment to `Case 1 - Aborted transactions does not contain the record batch from offsets 6-7 with producer id 2.` I added this item `new FetchResponseData.AbortedTransaction().setFirstOffset(6).setProducerId(1)` to `abortedTransactions` to show that even if offsets 6-7 is mentioned under `abortedTransactions`, they won't get filtered out  for archiving since they have a different `producerId`.

junrao commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/ShareConsumerTest.java:
```diff
@@ -2090,6 +2091,307 @@ public void testComplexShareConsumer() throws Exception {
+                // First transaction is committed.
+                produceCommittedTransaction(transactionalProducer, "Message 1");
+
+                ConsumerRecords<byte[], byte[]> records = waitedPoll(shareConsumer, 2500L, 1);
```
Does this force the test to wait for 2.5 secs? It's kind of long.

adixitconfluent replied:
I don't think that this will force the test to wait for 2.5 sec. This is the maximum time to block. The method returns immediately if there are records available. Otherwise, it will await the passed timeout.

junrao commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/ShareConsumerTest.java:
```diff
@@ -2090,6 +2091,307 @@ public void testComplexShareConsumer() throws Exception {
+                shareConsumer.commitSync();
+
+                // Precautionary poll so that the aborted transaction for Message 4 is fetched by the consumer.
+                shareConsumer.poll(Duration.ofMillis(5000));
```
Could we do `waitUntil` to avoid having to wait for 5 secs?

adixitconfluent replied:
I am not sure if I follow the comment. Since this is a poll to make sure aborted transaction for message 4 is fetched by the consumer, it won't return m any records. So what condition shall I put for `waitForCondition`? Anyways, I have reduced the poll duration to 2.5 seconds since I see that value is being used in polls for all the other test cases in this class.

junrao replied:
> Since this is a poll to make sure aborted transaction for message 4 is fetched by the consumer, it won't return m any records.

Since the `poll()` call doesn't return any record, we are forced to wait for 2.5 secs, right? Also, how do we know for sure that message 4 is fetched by the consumer?

adixitconfluent replied:
You are right, we would be are forced to wait for 2.5 secs. I have changed the testing for this scenario by setting an acknowledgement commit callback to verify that aborted marker offset for Message 4 (7L) is fetched and acknowledged by the consumer. I have used `TestUtils.waitForCondition()` and done a poll for `500ms` for the same. This will verify that transaction corresponding to message 4 is fetched by the consumer and also takes much less time now.

junrao commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -1154,7 +1163,8 @@ public void testAcquireWithMaxFetchRecords() {
             DEFAULT_FETCH_OFFSET,
-            fetchPartitionData(records)),
+            fetchPartitionData(records),
+                FETCH_ISOLATION_HWM),
```
indentation

junrao reviewed (commented):
@adixitconfluent : Thanks for the updated PR. A few more comments.

adixitconfluent committed:
Addressed Jun's comments round 4

adixitconfluent committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-16729 to resolve the merge conflicts

junrao commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -6669,15 +6691,441 @@ private String assertionFailedMessage(SharePartition sharePartition, Map<Long, L
+        assertEquals(16, actual.get(2).lastOffset());
+        assertEquals(1, actual.get(2).producerId());
+
+        // Case 2 - Aborted transactions contains the record batch from 4-5 with producer id 2.
```
Should we fix this comment too?

junrao commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/ShareConsumerTest.java:
```diff
@@ -2090,6 +2091,307 @@ public void testComplexShareConsumer() throws Exception {
+                produceAbortedTransaction(transactionalProducer, "Message 2");
+
+                // We will not receive any records since the transaction was aborted.
+                records = waitedPoll(shareConsumer, 2500L, 0);
```
Does this force the test to wait for 2.5 secs since it reads 0 record?

adixitconfluent replied:
yes, in this implementation it did. I have changed the testing implementation by setting an acknowledgement commit callback to verify that aborted marker offset for Message 2 (3L) is fetched and acknowledged by the consumer. I have used `TestUtils.waitForCondition()` and done a poll for `500ms` for the same. This will verify that transaction corresponding to message 4 is fetched by the consumer and also takes much less time now.

junrao reviewed (commented):
@adixitconfluent : Thanks for the updated PR. A couple of followup comments.

adixitconfluent committed:
Addressed Jun's comments round 5

junrao commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/ShareConsumerTest.java:
```diff
@@ -2090,6 +2091,339 @@ public void testComplexShareConsumer() throws Exception {
+                // aborted marker offset for Message 2 (3L) to be fetched and acknowledged by the consumer.
+                TestUtils.waitForCondition(() -> {
+                    ConsumerRecords<byte[], byte[]> pollRecords = shareConsumer.poll(Duration.ofMillis(500));
+                    return pollRecords.count() == 0 && partitionExceptionMap.containsKey(tp) && partitionOffsetsMap.containsKey(tp) && partitionOffsetsMap.get(tp).contains(3L);
```
Hmm, why do we wait for an exception? The exception is actually null. 

Also, is it true that control records will also be acknowledged?

adixitconfluent replied:
Yeah, we don't need to wait for an exception, so removed that check.
Yes, it is true that control records will be acknowledged with the type `ARCHIVE` which means `REJECT` or `GAP`. The client has the required functionality to handle control records.

junrao commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/ShareConsumerTest.java:
```diff
@@ -2090,6 +2091,339 @@ public void testComplexShareConsumer() throws Exception {
+             ShareConsumer<byte[], byte[]> shareConsumer = createShareConsumer("group1")) {
+            produceCommittedAndAbortedTransactionsInInterval(transactionalProducer, 10, 5);
+            shareConsumer.subscribe(Set.of(tp.topic()));
+            ConsumerRecords<byte[], byte[]> records = waitedPoll(shareConsumer, 2500L, 8);
```
Since we are not waiting for the timeout in the common case, it's better to use a longer timeout like 5secs to avoid flakiness.

junrao reviewed (commented):
@adixitconfluent : Thanks for the updated PR. A couple of more comments.

adixitconfluent committed:
Addressed Jun's comments round 6

junrao commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/ShareConsumerTest.java:
```diff
@@ -2090,6 +2091,335 @@ public void testComplexShareConsumer() throws Exception {
+
+                // Setting the acknowledgement commit callback to verify acknowledgement completion.
+                Map<TopicPartition, Set<Long>> partitionOffsetsMap = new HashMap<>();
+                shareConsumer.setAcknowledgementCommitCallback(new TestableAcknowledgementCommitCallback(partitionOffsetsMap, new HashMap<>()));
```
Could we fix `TestableAcknowledgementCommitCallback.onComplete `so that it only populates `partitionExceptionMap` is `exception` is not null?

junrao reviewed (commented):
@adixitconfluent : Thanks for the updated PR. Just one more comment.

adixitconfluent committed:
Addressed Jun's comment round 7

adixitconfluent committed:
Minor refactor to added comment addressed in Jun's review round 7

junrao reviewed (approved):
@adixitconfluent : Thanks for the updated PR. LGTM. I will wait to see if others have additional comments.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -2475,6 +2496,192 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) {
+    ) {
+        List<AcquiredRecords> result = new ArrayList<>();
+        Iterator<AcquiredRecords> acquiredRecordsListIter = acquiredRecordsList.iterator();
+        Iterator<RecordBatch> batchesToArchiveIterator = batchesToArchive.iterator();
+        if (!batchesToArchiveIterator.hasNext())
+            return acquiredRecordsList;
```
This should be the first check in the method, prior initializing `result = new ArrayList()`.

apoorvmittal10 reviewed (approved):
Thanks for the changes, LGTM. A minor comment.

adixitconfluent committed:
Addressed Apoorv's review comment
